lib/src/Uncertainty/Algorithm/MetaModel/Kriging/GeneralLinearModelAlgorithm.cxx

```diff
@@ -558,15 +564,37 @@ Scalar GeneralLinearModelAlgorithm::maximizeReducedLogLikelihood()
     return initialReducedLogLikelihood;
   }
   // At this point we have an optimization problem to solve
-  // Define the optimization problem
+  // Define the optimization problem. It is passed to solver_ attribute
+  // in order to have access to the locally built problem outside of
+  // the method, but it will not be the actual solver used for the
+  // optimization as it may be replaced by a multi-start version
   OptimizationProblem problem(reducedLogLikelihoodFunction);
   problem.setMinimization(false);
   problem.setBounds(optimizationBounds_);
   solver_.setStartingPoint(initialParameters);
   solver_.setProblem(problem);
-  LOGINFO(OSS(false) << "Solve problem=" << problem << " using solver=" << solver_);
-  solver_.run();
-  const OptimizationAlgorithm::Result result(solver_.getResult());
+  OptimizationAlgorithm actualSolver(solver_);
```

Comment:
It was my first idea, but I had to solve these issues:
+ I need to know the bounds to build the starting points, so the bounds have to be initialized before to build the MultiStart instance->easy to solve
+ What to do if the solver is modified by a call to its accessor? The simplest decision may be to change solver_ regardless of the value of ResourceMap::GetAsBool("GeneralLinearModelAlgorithm-UseMultiStart") . This way, I don't have to bother with the fact that the new solver may also be a MultiStart instance. Another decision could be to check if ResourceMap::GetAsBool("GeneralLinearModelAlgorithm-UseMultiStart") is true, in which case we know that solver_ is a MultiStart instance and we update its internal solver. But with this imperative view of ResourceMap::GetAsBool("GeneralLinearModelAlgorithm-UseMultiStart") it becomes difficult to switch to a non-MultiStart algorithm. You first have to change the value of this key (default is True), then you call the accessor.
+ What to do if the bounds are changed? If solver is not an instance of MultiStart there is nothing to change in solver_, but if it is an instance of MultiStart I have to check if ResourceMap::GetAsBool("GeneralLinearModelAlgorithm-UseMultiStart") is true, in which case GLMAlgo is in charge of the starting points, or if it is false in which case the user provided its own instance of MultiStart and it remains unchanged

Facing all these choices I preferred to delay the creation of the MultiStart instance up to the actual resolution of the optimization problem, where all the potential modifications of the bounds or the default solver have been made. If the user want to change the default solver and benefit from the automatic MultiStart strategy, he do it through a call to setOptimizationAlgorithm() with a solver able to solve an optimization problem with bounds. If he want to build his own MultiStart strategy, he just have to call setOptimizationAlgorithm() with the MultiStart instance of his choice: no need to change the ResourceMap key value.

With this details, do you agree with the current implementation?

Reply:
Thanks Régis for this detailed explanation of the difficulties involved. With this information, I think that your implementation is sound but should be moved to the `MultiStart` class itself. Commit 62f0a70 on my new branch [multistart](https://github.com/josephmure/openturns/tree/multistart) shows what I have in mind. The idea is:

- to have a new `MultiStart` constructor that takes the number of starting points as input
- to have the starting points generated the way you propose every time `MultiStart::setProblem` is called.

In order not to interfere with the workflow that involves manually setting the starting points, the new `MultiStart::selectStartingPoints` method checks that the attribute `MultiStart::nbSelectedStartingPoints_` is >0 and does nothing if it is not. All `MultiStart` constructors but the new set this attribute to 0.

Reply:
It is a good idea, however I am not 100% sure it works in all cases, and I don't see how to use it in my context in order to avoid writing code into run(), setOptimizationBounds() and setOptimizationAlgorithm(). Can you illustrate the use of the new constructor in the specific context of GLMAlgo? BTW, there is no check made on the base solver in MultiStart (not your fault of course). IMO we should exclude the possibility to have a MultiStart base solver (i.e. no recursive construction of MultiStart algorithms). If you extend MultiStart, why not allowing for a constructor based on a WeightedExperiment? This way, you can choose another sampling algorithm (LHS, other low discrepancy sequence...)

Reply:
Please check out PR #1797 for an updated version of my proposal. I will post an example of how you could use the new constructor within `GeneralLinearModelAlgorithm` as a GitHub "suggestion" . Of course, since it depends on PR #1797, you won't be able to accept it.